base/reflection.jl

```diff
@@ -1139,6 +1139,45 @@ function return_types(@nospecialize(f), @nospecialize(types=Tuple), interp=Core.
     return rt
 end
 
+"""
+    print_statement_costs(io::IO, f, types)
+
+Print type-inferred and optimized code for `f` given argument types `types`,
+prepending each line with its cost as estimated by the compiler's inlining engine.
+"""
+function print_statement_costs(io::IO, @nospecialize(f), @nospecialize(t); kwargs...)
```

Comment:
Maybe it makes sense to add a `@code_` version of this to InteractiveUtils. Could always be done later of course.

Reply:
No way, is that `@code_costs`?:joy:

Reply:
Or an argument to one of the others (`cost = true`)

Reply:
FYI, I'm providing [`@code_costs`](https://github.com/kimikage/CodeCosts.jl) and it's exactly @timholy who gave me the motivation to do so. :smile:

I have some objections about the display style, but I agree that Julia supports that feature.

Reply:
Oh, nice!

Reply:
I like your printing quite a lot better. But if you do it as separate code-display (`print(io, src); lines = readlines(io)`) then one risks having the two get misaligned, whereas this version ensures that the two will always stay aligned.

Demo:
```julia
julia> function buildexpr()
           items = [7, 3]
           ex = quote
               X = $items
               for x in X
                   println(x)
               end
           end
           return ex
       end
buildexpr (generic function with 1 method)

julia> @code_typed buildexpr()
CodeInfo(
1 ─ %1  = Core.tuple(7, 3)::Core.Const((7, 3), false)
│   %2  = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Vector{Int64}, svec(Any, Int64), 0, :(:ccall), Vector{Int64}, 2, 2))::Vector{Int64}
│         Base.arraysize(%2, 1)::Int64
└──       goto #7 if not true
2 ┄ %5  = φ (#1 => 1, #6 => %10)::Int64
│   %6  = φ (#1 => 1, #6 => %16)::Int64
│   %7  = φ (#1 => 1, #6 => %17)::Int64
│   %8  = Base.getfield(%1, %6, true)::Int64
│         Base.arrayset(false, %2, %8, %5)::Vector{Int64}
│   %10 = Base.add_int(%5, 1)::Int64
│   %11 = (%7 === 2)::Bool
└──       goto #4 if not %11
3 ─       goto #5
4 ─ %14 = Base.add_int(%7, 1)::Int64
└──       goto #5
5 ┄ %16 = φ (#4 => %14)::Int64
│   %17 = φ (#4 => %14)::Int64
│   %18 = φ (#3 => true, #4 => false)::Bool
│   %19 = Base.not_int(%18)::Bool
└──       goto #7 if not %19
6 ─       goto #2
7 ┄       goto #8
8 ─ %23 = Core._expr(:(=), :X, %2)::Expr
│   %24 = $(Expr(:copyast, :($(QuoteNode(:(for x = X
      #= REPL[4]:6 =#
      println(x)
  end))))))::Expr
│   %25 = Core._expr(:block, $(QuoteNode(:(#= REPL[4]:4 =#))), %23, $(QuoteNode(:(#= REPL[4]:5 =#))), %24)::Expr
└──       return %25
) => Expr
```
You can see statement `%24` took more than one line to print.

Reply:
I just made a change that will make it easier to fix that problem. Perhaps that will be fixed in a few hours. :smiley:

Reply:
I switched the order of the printing so the cost comes first...I just needed to look a little harder at how these functions were working. I edited the top post to show the current output.

Reply:
FYI, the unicode characters used in the block brackets may have different widths from the space (U+0020), even if the font is designed as "so-called" monospaced. :scream:

![alignment](https://user-images.githubusercontent.com/12679384/91778361-295f0d00-ec2d-11ea-8539-f4cba6a26f3c.png)

That' s not just a display issue. For example, when you want to extract an array of the cost numbers from the text, it gets annoying if they are inside the bracket.

So, I like the modified style :heart: